### PR TITLE
Hackbot check in reports

### PR DESCRIPTION
**Pull request by paked**
This PR adds some scripts which make my job on Monday morning of generating reports from the weekend's check ins super easy.

~Think of the current state as a WIP. I'm pushing this now in the hope of a review, and will come back in the morning after some sleep to finish it off.~ Ready for review as of 14/04/17

### CODE REVIEW

File: app/services/check_in_report_service.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      .map(&:leader)
+  end
+
+  def leaders_who_did_not_respond
+    @interactions
+      .where("data->>'failed_to_complete' = 'true'")
+      .map(&:leader)
+  end
+
+  def leaders_who_want_to_die
```
Comment (MaxWofford) [inline app/services/check_in_report_service.rb]: That's quite a name.

Comment (paked) [inline app/services/check_in_report_service.rb]: It's the taxonomy we're using... and I love it.

**Comment by paked** [comment]
@MaxWofford: ready for review

File: app/models/hackbot/interactions/text_conversation.rb [hunk 2]
```diff
@@ -34,8 +34,27 @@ def file_to_channel(filename, file)
...
+          dm_channel = channels.find { |im| im[:id] == data['channel'] }
+          dm_channel ? dm_channel[:user] : nil
+        end
+      end
+
       private
 
+      def channels
+        key = "im.list##{team.bot_access_token}"
+        Rails.cache.fetch(key, expires_in: 2.minutes) do
```
Comment (MaxWofford) [inline app/models/hackbot/interactions/text_conversation.rb]: Why are we caching here?

File: app/models/hackbot/interactions/text_conversation.rb [hunk 3]
```diff
@@ -34,8 +34,27 @@ def file_to_channel(filename, file)
...
+          # context.
+          dm_channel = channels.find { |im| im[:id] == data['channel'] }
+          dm_channel ? dm_channel[:user] : nil
+        end
+      end
+
       private
 
+      def channels
+        key = "im.list##{team.bot_access_token}"
```
Comment (MaxWofford) [inline app/models/hackbot/interactions/text_conversation.rb]: Could you find something else to use as the key here (as talked about in-person.

File: app/services/check_in_report_service.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    throw(Exception, "Need to specify a 'day' or a 'date'") if day_of.nil?
+
+    day_of = day_of.utc
+
+    min_day = day_of - 1.day
+    max_day = day_of + 1.day
+
+    @interactions = Hackbot::Interactions::CheckIn
+                    .where(state: 'finish')
+                    .where("created_at >= '#{min_day}' AND "\
```
Comment (MaxWofford) [inline app/services/check_in_report_service.rb]:
```diff
-                    .where("created_at >= '#{min_day}' AND "\
-             "created_at <= '#{max_day}'")
+                    .where("created_at >= '?'", min_day)
+                    .where("created_at <= '?'", max_day)
```

**Comment by MaxWofford** [comment]
Feel free to reassign me after addressing the stuff we talked about in-person.

File: app/services/check_in_report_service.rb (hunk 4 above)
Comment (paked) [inline app/services/check_in_report_service.rb]: @MaxWofford: What's the precedent for this?

**Comment by paked** [comment]
@MaxWofford: This is ready for another review/merge.

File: app/services/check_in_report_service.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      .map(&:leader)
+  end
+
+  def leaders_who_had_a_meeting
+    @interactions
+      .where("data->>'attendance' IS NOT NULL")
+      .map(&:leader)
+  end
+
+  def leaders_who_did_not_not_have_a_meeting
```
Comment (MaxWofford) [inline app/services/check_in_report_service.rb]: `leaders_who_did_*not*_not_have_a_meeting`

**Pull request merged**